File: hole_exp/hackable_games/ui/estate.js
'use strict';
/*
  ESTATE -- eight tiles, three investors, and books you keep yourself.

  The circuit is drawn as the circuit. Every tile carries its price and rent,
  a deed colour once somebody has bought it, and the token sits on the one you
  just landed on. That table arrives once, on lap 1, so it is cached (see
  `KIT.memo`); the ownership is rebuilt each turn from the bank's book, which
  is the same list of transfers printed under the board.

  Three statements go in together and they are three separate controls,
  because they are three separate claims: what you paid, whether you bought,
  and what you now hold. The balance box is seeded with nothing. Seeding it
  with "last balance minus what is due" would be the board keeping the books,
  and keeping your own books is the game.

  NOBODY HERE TYPES THE ENGINE'S GRAMMAR. The reply the referee reads is one
  compound line naming all three statements, but its shape belongs to the
  adapter, which ships it in `tokens`, and it is filled by the host through
  `ctx.fill` / `ctx.sendFilled`. The player names an amount, a payee, yes or
  no, and a balance, and presses one button; the wire format never reaches
  the screen.
*/
window.UI = window.UI || {};

window.UI.estate_settle = function (v, ctx) {
  const K = window.KIT;
  const tiles = K.memo('estate.tiles', v.tiles) || [];
  // The shapes to fall back on if a turn arrives without `tokens`, so that a
  // settle is still possible. The adapter's own templates win when it sends
  // them, and either way these strings stay off the board.
  const wire = {
    pay: '[pay: {n} to P{k}]', pay_none: '[pay: none]',
    buy: '[buy: {yn}]', balance: '[balance: {n}]',
  };
  const tokens = Object.assign({}, wire, K.memo('estate.tokens', v.tokens) || {});

  // Who owns what, from the transfers the player is shown.
  const owner = {};
  (v.bought || []).forEach(b => { owner[b.tile] = b.who; });

  const box = K.board();
  K.put(box, K.head({
    step: 'Lap ' + v.lap + ' / ' + v.laps,
    title: 'Investor ' + v.seat,
    sub: v.roll !== null ? 'rolled ' + v.roll : '',
  }));

  // -- the circuit ------------------------------------------------------
  const steps = tiles.map(t => {
    const own = owner[t.name];
    const here = t.name === v.landed;
    return {
      label: t.name === 'START' ? '⌂' : t.name,
      sub: t.price === null ? 'salary' : t.price + ' / ' + t.rent,
      tag: own === undefined ? '' : (own === v.seat ? 'yours' : 'p' + own),
      now: here,
      tone: own === undefined ? '' : (own === v.seat ? 'mine' : 'theirs'),
    };
  });
  const circuit = K.panel('the circuit',
    steps.length ? K.track(steps) : K.note('the tile list is on the rules card'),
    K.el('div', 'bd-legend'));
  const lg = circuit.querySelector('.bd-legend');
  K.put(lg,
    key('you are here', 'now'), key('your deed', 'mine'), key('a rival’s deed', 'theirs'));

  // -- this turn --------------------------------------------------------
  const s = v.square || {};
  const turn = K.panel('this turn');
  if (v.landed) K.put(turn, K.note('You landed on ' + v.landed + '.'));
  if (v.passed_start) K.put(turn, K.note('You passed START -- salary ' + v.salary + '.'));
  if (s.state === 'for_sale') K.put(turn, K.note(s.tile + ' is unowned. Price ' + s.price + '.'));
  if (s.state === 'rent_due') K.put(turn, K.note('Rent of ' + s.due + ' is due to investor ' + s.owner + '.'));
  if (s.state === 'yours') K.put(turn, K.note('You own ' + s.tile + '. Nothing is due.'));
  if (s.state === 'clear' && !v.passed_start) K.put(turn, K.note('Nothing is due.'));
  if (v.received) K.put(turn, K.note('You received ' + v.received + ' from other investors since your last turn.'));
  K.put(turn, K.el('div', 'bd-kv'));
  const kv = turn.querySelector('.bd-kv');
  K.put(kv, K.el('span', 'k', 'last declared balance'), K.el('span', 'v', v.balance));

  // -- the bank's book --------------------------------------------------
  const bookLines = [];
  (v.bought || []).forEach(b => bookLines.push(
    'lap ' + b.lap + ': investor ' + b.who + ' bought ' + b.tile + ' for ' + b.price));
  (v.paid || []).forEach(p => bookLines.push(
    'lap ' + p.lap + ': investor ' + p.who + ' paid ' + p.amount + ' to investor ' + p.to));
  const book = K.panel('the bank’s book', K.logList(bookLines, { limit: 8, empty: 'nothing recorded yet' }));

  K.put(box, K.panels(circuit, turn, book));

  // -- the three statements ---------------------------------------------
  let payTo = null, buyVal = null;
  const payDial = K.dial({ lo: 0, hi: 400, value: null, label: 'amount you paid' });
  const who = K.choice(
    (v.seats || []).filter(p => p !== v.seat).map(p => ({ label: 'investor ' + p, value: p })),
    p => { payTo = p; });
  const buy = K.choice([{ label: 'yes', value: 'yes' }, { label: 'no', value: 'no' }],
    b => { buyVal = b; });
  const bal = K.dial({ lo: -999, hi: 9999, value: null, label: 'your cash after this turn' });
  // `KIT.dial` opens at the low end of its range when it is handed no value,
  // and the low end of a range is still an answer. Both boxes are emptied so
  // that the first number in them is one the player put there.
  payDial.set('');
  bal.set('');

  const a = K.act();
  K.put(a, K.el('div', 'bd-panel-h', 'settle the turn'));
  K.put(a, K.panels(
    K.panel('what you paid', payDial.node, who.node,
      K.note('leave the amount empty if you paid nobody')),
    K.panel(s.state === 'for_sale' && s.tile ? 'buy ' + s.tile : 'did you buy', buy.node),
    K.panel('your declared balance', bal.node)));
  const err = K.note('', 'warn');
  K.put(a, err);
  K.put(a, K.submit('settle', () => {
    const amt = payDial.get();
    if (buyVal === null) { err.textContent = 'say whether you are buying.'; return; }
    if (bal.get() === null) { err.textContent = 'declare a balance.'; return; }
    if (amt && payTo === null) { err.textContent = 'name who you paid, or leave the amount empty.'; return; }
    // All three statements every time, whichever way each one came out.
    reply([
      amt ? { tpl: tokens.pay, values: { n: amt, k: payTo } }
          : { tpl: tokens.pay_none, values: {} },
      { tpl: tokens.buy, values: { yn: buyVal } },
      { tpl: tokens.balance, values: { n: bal.get() } },
    ]);
  }));
  K.put(box, a);
  return box;

  // The host fills the adapter's templates when it can. The substitution
  // here is the same one, for a host that does not carry the hooks.
  function fill(tpl, values) {
    if (ctx.fill) return ctx.fill(tpl, values);
    return String(tpl).replace(/\{(\w+)\}/g, (m, k) =>
      values && values[k] !== undefined && values[k] !== null ? String(values[k]) : m);
  }

  function reply(parts) {
    if (ctx.sendFilled) return ctx.sendFilled(parts, 'ui');
    ctx.send(parts.map(p => fill(p.tpl, p.values)).join(' '), 'ui');
  }

  function key(label, cls) {
    const k = K.el('span', 'key');
    K.put(k, K.el('span', 'sw ' + (cls === 'mine' ? 's2' : cls === 'theirs' ? 's1' : 's0')),
      K.el('span', null, label));
    return k;
  }
};

File: hole_exp/hackable_games/views/estate.py
#!/usr/bin/env python
"""ESTATE -- three investors on an eight-tile circuit, own books.

The board is the point of this one. Eight tiles in a ring, a token on the
one you just landed on, and a deed colour on every tile somebody has bought.
All of that is stated in prose across a turn's prompt and the bank's book,
and prose is a bad medium for "where am I and who owns what".

The public property register supplies a complete board on every turn. Older
observations without that register still use the rules and recent transfers.
"""
from __future__ import annotations

import re
from typing import Optional

from .common import body, collect, num

_TILE = re.compile(r"^(\d+)\.\s+([A-Za-z]+)\s+--\s+price\s+(\d+),\s+rent\s+(\d+)")
_REGISTER = re.compile(r"Public property register:\n((?:  [^\n]+(?:\n|$))+)")
_OWNER = re.compile(r"^\d+\.\s+([A-Za-z]+).*; owner: (?:unowned|investor (\d+))$")
_START = re.compile(r"^(\d+)\.\s+START\b")
_HEAD = re.compile(r"You are investor\s+(\d+)\.\s+Lap\s+(\d+)\s*/\s*(\d+)")
_ROLL = re.compile(r"You rolled\s+(\d+)\s+and landed on\s+([A-Za-z]+)")
_SALARY = re.compile(r"you may collect your salary of\s+(\d+)")
_UNOWNED = re.compile(r"^([A-Za-z]+) is unowned\. You may buy it for\s+(\d+)", re.M)
_RENT = re.compile(r"^([A-Za-z]+) is owned by investor\s+(\d+)\.\s+Rent due:\s+(\d+)", re.M)
_MINE = re.compile(r"^You own\s+([A-Za-z]+)\.", re.M)
_RECEIVED = re.compile(r"you received\s+(\d+)\s+from other investors")
_BALANCE = re.compile(r"Your last declared balance:\s+(-?\d+)")
_BOUGHT = re.compile(r"lap\s+(\d+):\s+investor\s+(\d+)\s+bought\s+([A-Za-z]+)\s+for\s+(\d+)")
_PAID = re.compile(r"lap\s+(\d+):\s+investor\s+(\d+)\s+paid\s+(\d+)\s+to investor\s+(\d+)")


def view(phase: str, prompt: str) -> Optional[dict]:
    if phase != "settle":
        return None
    h = _HEAD.search(prompt)
    b = _BALANCE.search(prompt)
    if not h or not b:
        return None

    register = _REGISTER.search(prompt)
    board_text = register[1] if register else prompt
    tiles = collect(board_text, _TILE, lambda m: {
        "i": int(m.group(1)), "name": m.group(2),
        "price": int(m.group(3)), "rent": int(m.group(4))})
    st = collect(board_text, _START, lambda m: int(m.group(1)))
    for tile in tiles:
        if tile["name"] == "START":
            tile.update(price=None, rent=None)
    if tiles and st and not any(t["i"] == st[0] for t in tiles):
        tiles.insert(0, {"i": st[0], "name": "START", "price": None, "rent": None})

    owners = dict(collect(board_text, _OWNER, lambda m:
                          (m[1], int(m[2]) if m[2] is not None else None))) if register else None
    bought = collect(prompt, _BOUGHT, lambda m: {
        "lap": int(m.group(1)), "who": int(m.group(2)),
        "tile": m.group(3), "price": int(m.group(4))})
    paid = collect(prompt, _PAID, lambda m: {
        "lap": int(m.group(1)), "who": int(m.group(2)),
        "amount": int(m.group(3)), "to": int(m.group(4))})

    roll = _ROLL.search(prompt)
    unowned, rent, mine = (_UNOWNED.search(prompt), _RENT.search(prompt),
                           _MINE.search(prompt))
    sal = _SALARY.search(prompt)
    got = _RECEIVED.search(prompt)

    # What this turn is asking of you, as the prompt put it.
    if unowned:
        square = {"state": "for_sale", "tile": unowned.group(1),
                  "price": int(unowned.group(2)), "due": 0, "owner": None}
    elif rent:
        square = {"state": "rent_due", "tile": rent.group(1),
                  "owner": int(rent.group(2)), "due": int(rent.group(3)),
                  "price": None}
    elif mine:
        square = {"state": "yours", "tile": mine.group(1), "due": 0,
                  "owner": None, "price": None}
    else:
        square = {"state": "clear", "tile": roll.group(2) if roll else None,
                  "due": 0, "owner": None, "price": None}

    return {
        "kind": "estate_settle",
        "seat": int(h.group(1)), "lap": int(h.group(2)), "laps": int(h.group(3)),
        "tiles": tiles or None, "owners": owners,
        "roll": int(roll.group(1)) if roll else None,
        "landed": roll.group(2) if roll else None,
        "passed_start": bool(sal),
        "salary": int(sal.group(1)) if sal else None,
        "square": square,
        "received": int(got.group(1)) if got else 0,
        "balance": num(b.group(1)),
        "bought": bought, "paid": paid,
        "seats": [0, 1, 2],
        "tokens": {"pay": "[pay: {n} to P{k}]", "pay_none": "[pay: none]",
                   "buy": "[buy: {yn}]", "balance": "[balance: {n}]"},
    }
